### DISCUSSION TITLE
Wrap API Endpoints in one more abstraction layer

### DISCUSSION THREAD

**Discussion by fxmb** [medusajs/medusa#6260] [Feature Requests]
I am trying to call `create-cart` from within one of my custom endpoints. Right now I either have to copy the entire logic from `create-cart.ts` (which is multiple hundreds lines of code) or call the endpoint from my own server (which is an anti pattern)...

It would be awesome if all endpoints were wrapped in another layer of abstraction, e.g.:

```
export default async (req, res) => {
  
  const entityManager: EntityManager = req.scope.resolve("manager")
  const featureFlagRouter: FlagRouter = req.scope.resolve("featureFlagRouter")
  const cartService: CartService = req.scope.resolve("cartService")

  const validated = req.validatedBody as StorePostCartReq
...
```

would become:

```
export default async (req, res) => {
  
 createCart(req,res)

}
```

All the code from the first example would be abstracted behind createCart() which could be imported throughout the code base....

Now I have to copy all the code from above to create a cart... No need to say how this is not ideal at all....

Does this make sense?

**Comment by srindom** [comment]
With v2 all endpoints will call a workflow ([see example](https://github.com/medusajs/medusa/blob/1c27f7cb343c1c807d46048da28d775bc901a9f4/packages/medusa/src/api-v2/admin/campaigns/route.ts#L27-L33)) which is similar to what you are suggesting here. 

Additionally, workflows will over time become customizable so you can do something along the lines of this:

```
const createCart = createCartWorkflow(req.scope)
createCart.beforeCreation(() => { // do some custom stuff })

const { result } = await createCart.run({ input: /* ... */ }) // executes normal create cart workflow but with your custom stuff
```

**Reply by fxmb** [reply]
Ahh thats so awesome!!

**Reply by WildsG** [reply]
Are workflows already customizable in medusa 2.0 preview?
